### ISSUE TITLE
Implement NullPropertyMappingStrategy to allow more flexibility when encountering null values

### ISSUE THREAD

**Issue by latonz** [riok/mapperly#98]
**Is your feature request related to a problem? Please describe.**
See https://github.com/riok/mapperly/discussions/96. Main use case is to merge two instances while only setting values which are non-null.

**Describe the solution you'd like**
Implement different strategies on how to handle null values when mapping property values.

**API**
```diff
public sealed class MapperAttribute : Attribute
{
+    [Obsolete($"Use {nameof(NullPropertyMappingStrategy)} instead")]
      public bool ThrowOnPropertyMappingNullMismatch
      {
          get =>  NullPropertyMappingStrategy == NullPropertyMappingStrategy.SetOrThrowIfNull;
          set =>  NullPropertyMappingStrategy = NullPropertyMappingStrategy.SetOrThrowIfNull;
       }

+    /// <summary>
+    /// Specifies the behaviour how to handle property mappings when <c>null</c> values are involved.
+    /// </summary>
+    public NullPropertyMappingStrategy NullPropertyMappingStrategy { get; set; } = NullPropertyMappingStrategy.SetOrIgnoreIfNull;
}

+/// <summary>
+/// Strategies on how to handle <c>null</c> values when mapping property values.
+/// </summary>
+public enum NullPropertyMappingStrategy
+{
+    /// <summary>
+    /// Sets the value if the target is nullable or the source is not <c>null</c>.
+    /// Ignores the mapping if the source is <c>null</c> and the target is not nullable
+    /// </summary>
+    SetOrIgnoreIfNull,
+
+    /// <summary>
+    /// Sets the value if the target is nullable or the source is not <c>null</c>.
+    /// Throws if the source is <c>null</c> and the target is not nullable
+    /// </summary>
+    SetOrThrowIfNull,
+
+    /// <summary>
+    /// Sets the value if the target is nullable or the source is not <c>null</c>.
+    /// Sets a default value otherwise (<c>default</c> for value types, empty string for strings, <c>new()</c> for classes; throw if no parameterless ctor exists).
+    /// </summary>
+    SetOrDefaultIfNull,
+
+    /// <summary>
+    /// Ignores the property if the source is <c>null</c>.
+    /// </summary>
+    IgnoreIfNull,
+
+    /// <summary>
+    /// Sets the value if the source is not <c>null</c>.
+    /// Sets a default value otherwise.
+    /// Similar to <see cref="SetOrDefaultIfNull"/> but does never set <c>null</c> values.
+    /// </summary>
+    DefaultIfNull,
+}
```

**Backward compatibility**
`SetOrIgnoreIfNull` is the same as `ThrowOnPropertyMappingNullMismatch=false`, `SetOrThrowIfNull` is the same as `ThrowOnPropertyMappingNullMismatch=true`. The existing `ThrowOnPropertyMappingNullMismatch` property is kept around but marked as obsolete. If `ThrowOnPropertyMappingNullMismatch` is `false`, the new `NullPropertyMappingStrategy` should be used. If `ThrowOnPropertyMappingNullMismatch` is `true`, always `SetOrThrowIfNull` should be used. The documentation should be updated accordingly.

**Additional context**
See https://github.com/riok/mapperly/discussions/96 and https://github.com/riok/mapperly/discussions/452.

**Comment by markushaslinger** [comment]
@latonz If I understand correctly this would allow handling of nullable value types (via `IgnoreIfNull`) while the current behavior is `SetOrDefaultIfNull`?

Is there currently (v2.7.0) a way to turn conversion of all nullable value types (int?, Foo?,...) to keep the null instead of creating a new default instance? Or do I have to create mapping methods for each Mapper?
Sorry if I missed something obvious in the documentation :)

**Comment by sunghwan2789** [comment]
> Is there currently (v2.7.0) a way to turn conversion of all nullable value types (int?, Foo?,...) to keep the null instead of creating a new default instance? Or do I have to create mapping methods for each Mapper? Sorry if I missed something obvious in the documentation :)

Defining a nullable mapper method allows you to keep null values:
```cs
public int? MapToNullableInt(int? source) => source;
```

**Comment by markushaslinger** [comment]
> Defining a nullable mapper method allows you to keep null values:
> 
> ```cs
> public int? MapToNullableInt(int? source) => source;
> ```

Yes, I know. My question was if that can be done globally or if I have to add it for every possible nullable type and to every Mapper class I define (I try to keep separate ones, one for each primary mapped type and related ones only). 
I cannot use a base class either, because the Mapper classes are static (to allow extension method syntax).

Tips for achieving a satisfying structure are appreciated!

**Comment by latonz** [comment]
@markushaslinger The current behaviour is equal to `SetOrIgnoreIfNull` of the proposed API (unless `ThrowOnPropertyMappingNullMismatch` or `ThrowOnMappingNullMismatch` is set). Therefore it should already work as you expect or I'm misunderstanding sth.
The current behaviour for properties when the source property is null and the target property is not nullable:

| Target property is required | ThrowOnPropertyMappingNullMismatch | Action                       |
|-----------------------------|------------------------------------|------------------------------|
| ✗                           | ✗                              | Ignored                      |
| ✗                           |  ✓                                | Throws ArgumentNullException |
| ✓                           | ✗                              | Default value assignment     |
| ✓                           |  ✓                                | Throws ArgumentNullException |

If I misunderstood sth. could you provide a complete example with a minimal reproducible of what you are getting and what you are expecting?

**Comment by markushaslinger** [comment]
> If I misunderstood sth. could you provide a complete example with a minimal reproducible of what you are getting and what you are expecting?

Pretty sure it's not a bug, but a feature request or me missing an option.
I'll provide a MWE to showcase the behavior (`default` assignment because target required, but source `null`).

```cs
using Riok.Mapperly.Abstractions;

var foo = new Foo(null, null);
var qux1 = foo.ToQux();
var qux2 = foo.ToQuxKeepNull();
Console.WriteLine(qux1); // null, 0 
Console.WriteLine(qux2); // null, null - desired behavior


public sealed record Foo(string? Bar, int? Baz);
public sealed record Qux(string? Bar, int? Baz);

[Mapper]
static partial class TestMapper
{
    public static partial Qux ToQux(this Foo foo);
}

[Mapper]
static partial class KeepNullMapper
{
    public static partial Qux ToQuxKeepNull(this Foo foo);

    private static int? MapWithNull(int? v) => v;
}
```

The generated code for `ToQux` is - as far as I can tell as currently implemented and expected - as follows (`default` assignment instead of keeping `null`):

```cs
public static partial Qux ToQux(this Foo foo)
{
    var target = new Qux(foo.Bar ?? default, foo.Baz ?? default);
    return target;
}
```

=> I'd like to configure it, so that it will keep the `null`s - not a problem for reference types, because the `default` is `null`, but that doesn't work for nullable value types.
Basically some easy way to get the `ToQuxKeepNull` behavior without having to add `MapWithNull`-like conversion methods everywhere.

If such an option exists please tell me how, otherwise I'd like to humbly ask for it :)

**Comment by latonz** [comment]
Ah got it, in the case you mentioned Mapperly is mapping to constructor parameters. The behaviour you mentioned is a Bug in v2.7.0 and should be resolved in `2.8.0-next.1`.

**Comment by markushaslinger** [comment]
Awesome, thanks for checking and the update!
I tested with `2.8.0-next.1` and can confirm it seems to be working as expected 👍

**Comment by terekhovkg** [comment]
It seems that the described behavior is preserved for the init properties

`
internal enum DomainEnum
{
    NonState = 1,
    State = 2,
    Federal = 3
}

enum ContractEnum
{
    NonState = 1,
    State = 2,
    Federal = 3
}

class Foo
{
    public DomainEnum? Enum { get; set; }
}

class FooDto
{
    public ContractEnum? Enum { get; init; }
}

[Mapper(EnumMappingStrategy = EnumMappingStrategy.ByName)]
static partial class FooMapper
{
    public static partial FooDto ToDto(this Foo foo);
}
`

The generated code for FooDto is

`
public static partial global::BG.Enrichment.App.Mappers.FooDto ToDto(this global::BG.Enrichment.App.Mappers.Foo foo)
{
    var target = new global::BG.Enrichment.App.Mappers.FooDto()
    {Enum = foo.Enum != null ? MapToContractEnum(foo.Enum.Value) : default};
    return target;
}
`

**Comment by terekhovkg** [comment]
It seems that the described behavior is preserved for the init properties

```C#
enum DomainEnum
{
    NonState = 1,
    State = 2,
    Federal = 3
}

enum ContractEnum
{
    NonState = 1,
    State = 2,
    Federal = 3
}

class Foo
{
    public DomainEnum? Enum { get; set; }
}

class FooDto
{
    public ContractEnum? Enum { get; init; }
}

[Mapper(EnumMappingStrategy = EnumMappingStrategy.ByName)]
static partial class FooMapper
{
    public static partial FooDto ToDto(this Foo foo);
}
```

The generated code for FooDto is

```C#
static partial global::BG.Enrichment.App.Mappers.FooDto ToDto(this global::BG.Enrichment.App.Mappers.Foo foo)
{
    var target = new global::BG.Enrichment.App.Mappers.FooDto()
    {Enum = foo.Enum != null ? MapToContractEnum(foo.Enum.Value) : default};
    return target;
}
```

**Comment by moxplod** [comment]
I have the exact use case for partial updates. I am using it for Importing via CSV or an api where a single field could be updated.

More details here: https://github.com/riok/mapperly/discussions/452 

From the above IgnoreIfNull would be the right use case for this, I believe.

**Comment by michelbieleveld** [comment]
Same here, also for partial updates, currently unusable unfortunately for my use-case and just will be writing manually 

if (propertyValue != null)
{
  property.Value = propertyValue
}

or something similar for all properties. Similar to Mapster's
```
TypeAdapterConfig<TSource, TDestination>
    .NewConfig()
    .IgnoreNullValues(true);
```
or conditional mapping in AutoMapper would solve this problem. The attribute above would serve. Not sure how complex to implement this is.

**Comment by wbuck** [comment]
So lets say I want to map properties from source to destination without `null` checks (unconditionally map the values), which option would I use? I assume it would be `DefaultIfNull`?

**Comment by latonz** [comment]
@wbuck depends on what you mean by "unconditionally". Mapperly does never assign `null` to a non-nullable property.
For example if the source property is `string?` and the target property is `string`, `null` will never be assigned directly. `DefaultIfNull` would assign an empty string if the source is `null`.

**Comment by wbuck** [comment]
In my particular case I'm mapping from a type which resides in a dotnet standard 2.0 project over to another type which resides in a dotnet 6.0 project. The types from the dotnet standard project can't use `nullable` reference types so the library assumes that all properties in the source could potentially be `null` (which is a good default behavior).

In this case the way to override this behavior would be to make the properties in target `nullable`.
I am curious as to what the difference between `SetOrDefaultIfNull` and `DefaultIfNull` is?
